FirebaseMLModelDownloader/Sources/ModelDownloadTask.swift

```diff
@@ -94,93 +81,96 @@ class ModelDownloadTask: NSObject {
-    // TODO: Handle waiting for connectivity, if needed.
-  }
-
-  func urlSession(_ session: URLSession, didBecomeInvalidWithError error: Error?) {
-    // TODO: Handle configuration errors.
-  }
-
-  /// Handle client-side errors.
-  func urlSession(_ session: URLSession, task: URLSessionTask, didCompleteWithError error: Error?) {
-    // TODO: Log this.
-    guard task == downloadTask else { return }
-    guard let error = error else { return }
-    /// Unable to resolve hostname or connect to host.
-    DispatchQueue.main.async {
-      self.downloadHandlers
-        .completion(.failure(.internalError(description: error.localizedDescription)))
+  func download() {
+    downloader.downloadFile(with: remoteModelInfo.downloadURL,
+                            progressHandler: { downloadedBytes, totalBytes in
+                              /// Fraction of model file downloaded.
+                              let calculatedProgress = Float(downloadedBytes) / Float(totalBytes)
+                              DispatchQueue.main.async {
+                                self.downloadHandlers.progressHandler?(calculatedProgress)
+                              }
+                            }) { result in
+      switch result {
+      case let .success(response):
+        self.handleResponse(response: response.urlResponse, tempURL: response.fileURL)
+      case let .failure(error):
+        var downloadError: DownloadError
+        switch error {
+        case let FileDownloaderError.networkError(error):
+          downloadError = .internalError(description: ModelDownloadTask
+            .ErrorDescription
+            .invalidHostName(error
+              .localizedDescription))
+        case let FileDownloaderError.sessionInvalidated(error):
+          downloadError = .failedPrecondition
+        case FileDownloaderError.unexpectedResponseType:
+          downloadError = .internalError(description: ModelDownloadTask
+            .ErrorDescription.invalidServerResponse)
+
+        default:
+          downloadError = .internalError(description: ModelDownloadTask.ErrorDescription
+            .unknownDownloadError)
+        }
+        DispatchQueue.main.async {
+          self.downloadHandlers
+            .completion(.failure(downloadError))
+        }
+      }
     }
   }
 
-  func urlSession(_ session: URLSession,
-                  downloadTask: URLSessionDownloadTask,
-                  didFinishDownloadingTo location: URL) {
-    // TODO: Log this.
-    guard downloadTask == self.downloadTask else { return }
-    guard let response = downloadTask.response as? HTTPURLResponse else {
-      DispatchQueue.main.async {
-        self.downloadHandlers
-          .completion(.failure(.internalError(description: ModelDownloadTask
-              .ErrorDescription.invalidServerResponse)))
-      }
+  /// Fetch model info again and retry download if allowed.
+  func maybeRetryDownload() {
```

Comment:
nit: in the future I would suggest to consider moving retry logic up to `ModelDownloader` level. In this case the `ModelDownloadTask` will just complete with an error and `modelInfoRetriever` dependency will not be required here.

Reply:
I agree. I'll create a PR to do this.